applications/TrilinosApplication/custom_strategies/builder_and_solvers/trilinos_block_builder_and_solver.h

```diff
@@ -4,954 +4,868 @@
+            for (typename DofsVectorType::iterator i = dof_list.begin();
+                 i != dof_list.end(); ++i)
+                temp_dofs_array.push_back(i->get());
         }
 
+        temp_dofs_array.Unique();
+        BaseType::mDofSet = temp_dofs_array;
 
-        Doftemp.Unique();
-
-        BaseType::mDofSet = Doftemp;
-
-        //throws an execption if there are no Degrees of freedom involved in the analysis
-        if (BaseType::mDofSet.size()==0)
+        // throws an exception if there are no Degrees of freedom involved in
+        // the analysis
+        if (BaseType::mDofSet.size() == 0)
             KRATOS_ERROR << "No degrees of freedom!";
 
-    // If reactions are to be calculated, we check if all the dofs have reactions defined
-    // This is tobe done only in debug mode
-
-    #ifdef KRATOS_DEBUG
-
-    if(BaseType::GetCalculateReactionsFlag())
-    {
-        for(auto dof_iterator = BaseType::mDofSet.begin(); dof_iterator != BaseType::mDofSet.end(); ++dof_iterator)
-        {
-                KRATOS_ERROR_IF_NOT(dof_iterator->HasReaction()) << "Reaction variable not set for the following : " <<std::endl
-                    << "Node : "<<dof_iterator->Id()<< std::endl
-                    << "Dof : "<<(*dof_iterator)<<std::endl<<"Not possible to calculate reactions."<<std::endl;
+#ifdef KRATOS_DEBUG
+        // If reactions are to be calculated, we check if all the dofs have
+        // reactions defined This is to be done only in debug mode
+        if (BaseType::GetCalculateReactionsFlag()) {
+            for (auto dof_iterator = BaseType::mDofSet.begin();
+                 dof_iterator != BaseType::mDofSet.end(); ++dof_iterator) {
+                KRATOS_ERROR_IF_NOT(dof_iterator->HasReaction())
+                    << "Reaction variable not set for the following : " << std::endl
+                    << "Node : " << dof_iterator->Id() << std::endl
+                    << "Dof : " << (*dof_iterator) << std::endl
+                    << "Not possible to calculate reactions." << std::endl;
+            }
         }
-    }
-    #endif
-
+#endif
         BaseType::mDofSetIsInitialized = true;
 
         KRATOS_CATCH("")
     }
 
-    //**************************************************************************
-    //**************************************************************************
-    void SetUpSystem(
-        ModelPart& r_model_part
-    ) override
+    /**
+     * @brief Organises the dofset in order to speed up the building phase
+     *          Sets equation id for degrees of freedom
+     * @param rModelPart The model part of the problem to solve
+     */
+    void SetUpSystem(ModelPart& rModelPart) override
     {
-
-        // Set equation id for degrees of freedom
-
         int free_size = 0;
-        //int fixed_size = 0;
-
-        int rank;
-        MPI_Comm_rank(MPI_COMM_WORLD,&rank);
+        auto& r_comm = rModelPart.GetCommunicator();
```

Comment:
```suggestion
        const auto& r_comm = rModelPart.GetCommunicator();
```

Reply:
That is not possible as there is `SynchronizeDofs` is called later on .. !

Reply:
Probably for a separate PR, but I believe SynchronizeDofs could/should be const...